fgpyo/fasta/builder.py

```diff
@@ -0,0 +1,231 @@
+
+
+class ContigBuilder:
+    """Builder for constructing new contigs, and adding bases to existing contigs.
+    Existing contigs cannot be overwritten, each contig name in FastaBuilder must
+    be unique. Assembly and species information can be overwritten at instantiation however
+    defaults are provided if 'assembly' an/or 'species' are not provided.
+
+    Attributes:
+        name: Unique contig ID, ie., "chr10"
+        assembly: Assembly information, if None default is 'testassembly'
+        species: Species information, if None default is 'testspecies'
+        bases:  The bases to be added to the contig ex "A"
+
+    """
+
+    def __init__(
+        self,
+        name: str,
+        assembly: str = "testassembly",
+        species: str = "testspecies",
+    ):
+        self.name = name
+        self.assembly = assembly
+        self.species = species
+        self.bases = ""
+
+    def add(self, bases: str, times: int) -> "ContigBuilder":
+        """
+        Method for adding bases to a new or existing instance of ContigBuilder.
+
+        Args:
+            bases: The bases to be added to the contig
+            times: The number of times the bases should be repeated
+
+        Example
+        add("AAA", 2) results in the following bases -> "AAAAAA"
+        """
+        # Remove any spaces in string and enforce upper case format
+        bases = bases.replace(" ", "").upper()
+        self.bases += str(bases * times)
+        return self
+
+
+class FastaBuilder:
+    """Builder for constructing sets of one or more contigs.
+
+    Provides the ability to manufacture sets of contigs from minimal input, and automatically
+    generates the information necessary for writing the FASAT file, index, and dictionary.
+
+    A builder is constructed from an assembly, species, and line length. All attributes have
+    defaults, however these can be overwritten.
+
+    Contigs are added to FastaBuilder using:
+    :func:`~fgpyo.fasta.builder.FastaBuilder.add`
+
+    Bases are added to existing contigs using:
+    :func:`~fgpyo.fasta.builder.FastaBuilder.add.add`
+
+    Once accumulated the contigs can be written to a file using:
+    :func:`~fgpyo.fasta.builder.FastaBuilder.to_file`
+
+    Calling to_file() will also generate the fasta index (.fai) and sequence dictionary (.dict).
+
+    Attributes:
+        assembly: Assembly information, if None default is 'testassembly'
+        species: Species, if None default is 'testspecies'
+        line_length: Desired line length, if None default is 80
+        contig_builders: Private dictionary of contig names and instances of ContigBuilder
+    """
+
+    def __init__(
+        self,
+        assembly: str = "testassembly",
+        species: str = "testspecies",
+        line_length: int = 80,
+    ):
+        self.assembly: str = assembly
+        self.species: str = species
+        self.line_length: int = line_length
+        self.__contig_builders: Dict[str, ContigBuilder] = {}
```

Comment:
Why two underscores?

Reply:
one underscore is a weak internal reference where sub-classes and others can reference it, two is "I am very serious this is private" and will get mangled.